### ISSUE TITLE
Suddenly stops working

### ISSUE THREAD

**Issue by marutha** [Eonblast/Emysql#45]
I have scenario where  my emysql application stops processing queries.. This does not happen always but it happens rarely. 

I did a netstat on the client machine and could see that socket was on CLOSE_WAIT state.

Once this state comes, all the queries suspends and nothing happens.

**Comment by gulige** [comment]
It maybe the reason that mysql server closed the socket for long time idle.
emysql has no active reconnect mechanism as soon as socket is closed, however it attempts to support passive reconnect when you use the bad socket in the connection pool again.
It seems some issue in code as I commented for Issue #28 below:

For the line 573 in emysql.erl:
573 {'DOWN', Mref, process, Pid, {_, closed}} ->
I doubt it could cover all tcp errors especially gen_tcp:send/2 in emysql_tcp.erl
Please see the description of the returns of gen_tcp:send/2 for reference.
I think all the other tcp error cases should be matched in this case branch for better experience.

**Comment by hdiedrich** [comment]
Hi Maruthavanan,

thanks for reporting! 

Gulige, you are suggesting as a remedy, that the client needs to react to closing from the server side and actively re-open? Would you propose a pull request for that?

If the client sockets get in CLOSE_WAIT, your are suggesting line 573 of emysql.erl, or following, should be able to catch that and re-open the socket of the connection?

As you are correctly pointing out, this was experienced before (#28) so it would be cool to get it addressed. Anyone having a proposal how to create the state without waiting 3 to 4 days? Probably as simple as reducing the MySQL server idle socket time out?

**Comment by gulige** [comment]
Hi Hdiedrich,

I modified a few lines(45/163/194/204) in emysql_tcp.erl to indicate tcp errors as below.
exit({failed_to_xxx, Reason}) => exit({tcp_error, failed_to_xxx, Reason})
Then changed line 573 of emysql.erl to:
{'DOWN', Mref, process, Pid, {tcp_error, _, _}} ->
So, it can capture all tcp errors.
Originally, {'DOWN', Mref, process, Pid, {_, closed}} ->
I think it can only capture receiving cases when socket in passive mode.

As for MySQL server idle socket time out, you can adjuest the wait_timeout variable in mysql server, it defaults to 8 hours

**Comment by gulige** [comment]
Hi Hdiedrich,

Further explanation :-)
For the two receive branchs in monitor_work/3:
573 {'DOWN', Mref, process, Pid, {_, closed}} ->
584 {'DOWN', Mref, process, Pid, Reason} ->
The former branch will reconnect and re-execute {M, F, A}, while the latter one just reconnect.
It's reasonable for the latter one to NOT re-execute {M, F, A}, because it makes no sense in the case of non-tcp-issue.
The point is the former one cannot capture all tcp error cases, the tcp sending error cases will follow the latter logic branch. We shouldn't let the code of using emysql to re-execute the failed query in such scenario. Even if re-try, it could fail again for unluckily being executed with another bad connection in the connection pool.

**Comment by hdiedrich** [comment]
Thanks a lot for the investigation!

I am not sure right now. What must not happen is that things get executed twice because the error occurred half-way through the send and the server was able to act on it. 

For that reason, I wonder if in this case changing the time out on the server might be the only safe solution. Or creating a ping from the client side, which should be pretty simple and robust.

Would you maybe like to propose a patch along the lines of what you are describing above via pull request?

Thanks,
Henning

**Comment by gulige** [comment]
Henning,

Yes, I agree with you. Ping is the only safe solution. Could emysql provide an interface to do it? Or could the controlling process of socket in emysql reconnect to mysql server as soon as socket was broken?

Thanks,

**Comment by hdiedrich** [comment]
I don't know what message to the server would be neutral and can be used as ping. Maybe one exist for that stay-alive purpose. I would think it should be implemented as a completely independent timer that triggers the socket controlling process to send a 'ping' message to the server at a fixed interval, to keep it alive. We just received a patch for better integration of options and flags in the add_pool call.

https://github.com/Eonblast/Emysql/pull/43/files

That can be used to set the frequency for the ping, down the road.

It could also be an option to let the socket controlling process re-connect as soon as it sees that the socket is being closed. I'd have to look into that. It might also break the logic.

**Comment by gulige** [comment]
Henning,

Based on the discussions above, I submitted a pull request. The way is to test a connection with ping and reset the connection if ping failed before further processing a query. Thus, I think it's safe with no things get executed twice.

https://github.com/Eonblast/Emysql/pull/51

As to re-connecting as soon as socket is being closed, actually, I think it's not a good method when mysql server is shutdown manually for somewhat long time and restart it later. It's better to attempt to re-connect when needed.

**Comment by Eonblast** [comment]
Thanks a lot for putting this together!

But it would send a ping before every single call sent to the server?

**Comment by gulige** [comment]
Yes, a ping will be sent per single call in my pull request. It may be put into getConnection(emysql_conn_mgr:wait_for_connection/1 and lock_connection/1) functions to verify sanity rather than in emysql:monitor_work/3. However, getConnection is bound with monitor_work by execute interfaces.

**Comment by gulige** [comment]
Maybe we can set a timer to periodically test all connections in pool. It will be somewhat complex to handle such scenario that it makes no sense to periodically test when mysql server is down.

**Comment by taybin** [comment]
You could have it check every N queries.

**Comment by gulige** [comment]
taybin,

Thanks for your comments. I used a similar way as you said. I will send out a new pull request later.

**Comment by gulige** [comment]
Henning, Taybin,
I just submited a new pull request and closed the old one.
The new one is here: https://github.com/Eonblast/Emysql/pull/54

I added an env variable called conn_test_period, which is used to judge if need to test connection, so as to avoid ping per single call.

conn_test_period should be set to a value that is a little less than the wait_timeout of mysql server.
However, there is still an issue hard for me to solve in the scenario that mysql server is shutdown. In this scenario,  emysql cannot know it because the socket in connection is set to passive mode. At this stage, there are two actions can be taken, one is to restart emysql, another is to set conn_test_period to zero (retrogress to each ping per single call).
